src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Check.java

```diff
@@ -3955,6 +3958,128 @@ private void checkCyclicConstructor(JCClassDecl tree, Symbol ctor,
         }
     }
 
+/* *************************************************************************
+ * Verify the proper placement of super()/this() calls.
+ *
+ *    - super()/this() may only appear in constructors
+ *    - There must be at most one super()/this() call per constructor
+ *    - The super()/this() call, if any, must be a top-level statement in the
+ *      constructor, i.e., not nested inside any other statement or block
+ *    - There must be no return statements prior to the super()/this() call
+ **************************************************************************/
+
+    void checkSuperInitCalls(JCClassDecl tree) {
+        new SuperThisChecker().check(tree);
+    }
+
+    private class SuperThisChecker extends TreeScanner {
```

Comment:
Thanks - fixed in 599d761b309.

Reply:
Actually, I disagree a bit with the direction here. Having more mutable state is, I believe, not great. In this case the state has to be discarded after a call to `check`, so IMHO having mutable state lying around is a smell. What do you think @vicente-romero-oracle ?

Reply:
yes @mcimadamore I see your point. And after my comments I think we discussed a similar issue in another bug and we went for the non-mutable state. After revisiting the code I agree that in this case it could be better to avoid having mutable state.

Reply:
I agree with you both, and will revert this back to creating a new instance each time. Thanks.

Reply:
Thanks, and sorry for being late to the party!